pkg/service/auth_context.go

```diff
@@ -47,71 +48,121 @@ func NewAuthContext(parentCtx context.Context, req *auth.CheckRequest, apiConfig
+	ctx, cancel := context.WithCancel(context.Background())
+	errGroup, errCtx := errgroup.WithContext(ctx)
 
 	for _, authConfig := range authConfigs {
 		objConfig := authConfig
 		errGroup.Go(func() error {
-			return authContext.evaluateAuthConfig(ctx, objConfig, cb)
+			return authContext.evaluateAuthConfig(errCtx, cancel, objConfig, cb)
 		})
 	}
 	if err := errGroup.Wait(); err != nil {
 		return err
 	} else {
-		authCxtLog.Info("Successfully fetched all auth objects.", "authConfigs", authConfigs)
+		authCtxLog.Info("Successfully evaluated all auth objects.", "authConfigs", authConfigs)
 		return nil
 	}
 }
 
-// Evaluate evaluates all steps of the auth pipeline (identity → metadata → policy enforcement)
-func (authContext *AuthContext) Evaluate() error {
-	// identity
-	if err := authContext.evaluateAuthConfigs(authContext.API.IdentityConfigs,
-		func(conf common.AuthConfigEvaluator, authObj interface{}) {
-			// Convert from interfaceType to SpecificType
-			v, _ := conf.(*config.IdentityConfig)
-			authCxtLog.Info("Identity", "Config", conf, "AuthObj", authObj)
-			authContext.Identity[v] = authObj
-		}); err != nil {
-		return err
+// EvaluateIdentity evaluates given identity configs and retrieves the first valid identity object.
+func (authContext *AuthContext) EvaluateIdentity() error {
+	identityConfigs := authContext.API.IdentityConfigs
+	if len(identityConfigs) <= 0 {
+		return errors.New("no identity configs found")
 	}
 
-	// metadata
-	if err := authContext.evaluateAuthConfigs(authContext.API.MetadataConfigs,
-		func(conf common.AuthConfigEvaluator, authObj interface{}) {
+	successChannel := make(chan bool)
+
+	go func() {
+		defer close(successChannel)
+		authContext.evaluateAuthConfigs(identityConfigs,
+			func(conf common.AuthConfigEvaluator, cancelCtx context.CancelFunc, authObj interface{}, err error) error {
+				if err != nil {
+					authCtxLog.Error(err, "error evaluating identity config", "config", conf)
+					return nil
+				}
+				// Convert from interfaceType to SpecificType
+				v, _ := conf.(*config.IdentityConfig)
+				authCtxLog.Info("Identity", "Config", conf, "AuthObj", authObj)
+				authContext.Identity[v] = authObj
+				cancelCtx()
+				successChannel <- true
+				return nil
+			})
+	}()
+	success := <-successChannel
+	if !success {
+		return errors.New("error evaluating identity configs")
+	}
+	return nil
+}
+
+// EvaluateMetadata checks evaluates the given metadata configs and retrieve possible info.
+func (authContext *AuthContext) EvaluateMetadata() {
+	metadataConfigs := authContext.API.MetadataConfigs
+	if len(metadataConfigs) <= 0 {
+		authCtxLog.Info("No metadata configs found", "Configs", metadataConfigs)
+		return
+	}
+	authContext.evaluateAuthConfigs(metadataConfigs,
+		func(conf common.AuthConfigEvaluator, cancelCtx context.CancelFunc, authObj interface{}, err error) error {
+			if err != nil {
+				authCtxLog.Error(err, "error evaluating metadata config", "config", conf)
+				return nil
```

Comment:
Same as https://github.com/3scale-labs/authorino/pull/41/files#r576698099

Reply:
yup, the same, it seems non blocking is kind of default with the new behaviour

Reply:
The "new" part is on me. It is something that was obvious to me from start but only actually realised that the code was not right (and it wasn't even before concurrency was introduced!) when I started working on https://github.com/3scale-labs/authorino/pull/46.